// trezor.js/lib/utils/signbjstx.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.signBjsTx = signBjsTx;

var _bchaddrjs = require('bchaddrjs');

var _bchaddrjs2 = _interopRequireDefault(_bchaddrjs);

var _utxoLib = require('@trezor/utxo-lib');

var bitcoin = _interopRequireWildcard(_utxoLib);

var _trezortypes = require('../trezortypes');

var trezor = _interopRequireWildcard(_trezortypes);

var _hdnode = require('./hdnode');

var hdnodeUtils = _interopRequireWildcard(_hdnode);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

bitcoin.Transaction.USE_STRING_VALUES = true;

// TODO refactor this using string types


function input2trezor(input, sequence) {
    var hash = input.hash,
        index = input.index,
        path = input.path,
        amount = input.amount;

    return {
        prev_index: index,
        prev_hash: reverseBuffer(hash).toString('hex'),
        address_n: path,
        script_type: input.segwit ? 'SPENDP2SHWITNESS' : 'SPENDADDRESS',
        amount: amount,
        sequence: sequence
    };
}

function _flow_makeArray(a) {
    if (!Array.isArray(a)) {
        throw new Error('Both address and path of an output cannot be null.');
    }
    var res = [];
    a.forEach(function (k) {
        if (typeof k === 'number') {
            res.push(k);
        }
    });
    return res;
}

function output2trezor(output, network, isCashaddress) {
    if (output.address == null) {
        if (output.opReturnData != null) {
            if (output.value != null) {
                throw new Error('Wrong type.');
            }

            var data = output.opReturnData;
            return {
                amount: '0',
                op_return_data: data.toString('hex'),
                script_type: 'PAYTOOPRETURN'
            };
        }

        if (!output.path) {
            throw new Error('Both address and path of an output cannot be null.');
        }

        var pathArr = _flow_makeArray(output.path);

        var _amount = output.value;
        if (output.segwit) {
            return {
                address_n: pathArr,
                amount: _amount,
                script_type: 'PAYTOP2SHWITNESS'
            };
        } else {
            return {
                address_n: pathArr,
                amount: _amount,
                script_type: 'PAYTOADDRESS'
            };
        }
    }
    var address = output.address;
    if (typeof address !== 'string') {
        throw new Error('Wrong type.');
    }

    // $FlowIssue
    var amount = output.value;

    isScriptHash(address, network, isCashaddress);

    // cashaddr hack, internally we work only with legacy addresses, but we output cashaddr
    return {
        address: isCashaddress ? _bchaddrjs2.default.toCashAddress(address) : address,
        amount: amount,
        script_type: 'PAYTOADDRESS'
    };
}

function signedTx2bjsTx(signedTx, network) {
    var res = bitcoin.Transaction.fromHex(signedTx.message.serialized.serialized_tx, network);
    return res;
}

function bjsTx2refTx(tx) {
    var extraData = tx.getExtraData();
    var version_group_id = bitcoin.coins.isZcash(tx.network) ? parseInt(tx.versionGroupId, 16) : null;
    return {
        lock_time: tx.locktime,
        version: tx.isDashSpecialTransaction() ? tx.version | tx.type << 16 : tx.version,
        hash: tx.getId(),
        inputs: tx.ins.map(function (input) {
            return {
                prev_index: input.index,
                sequence: input.sequence,
                prev_hash: reverseBuffer(input.hash).toString('hex'),
                script_sig: input.script.toString('hex')
            };
        }),
        bin_outputs: tx.outs.map(function (output) {
            return {
                amount: typeof output.value === 'number' ? output.value.toString() : output.value,
                script_pubkey: output.script.toString('hex')
            };
        }),
        extra_data: extraData ? extraData.toString('hex') : null,
        version_group_id: version_group_id
    };
}

function _flow_getPathOrAddress(output) {
    if (output.path) {
        var _path = output.path;
        return _flow_makeArray(_path);
    }
    if (typeof output.address === 'string') {
        return output.address;
    }
    throw new Error('Wrong output type.');
}

function _flow_getSegwit(output) {
    if (output.segwit) {
        return true;
    }
    return false;
}

function deriveWitnessOutput(pkh) {
    // see https://github.com/bitcoin/bips/blob/master/bip-0049.mediawiki
    // address derivation + test vectors
    var scriptSig = new Buffer(pkh.length + 2);
    scriptSig[0] = 0;
    scriptSig[1] = 0x14;
    pkh.copy(scriptSig, 2);
    var addressBytes = bitcoin.crypto.hash160(scriptSig);
    var scriptPubKey = new Buffer(23);
    scriptPubKey[0] = 0xa9;
    scriptPubKey[1] = 0x14;
    scriptPubKey[22] = 0x87;
    addressBytes.copy(scriptPubKey, 2);
    return scriptPubKey;
}

function deriveOutputScript(pathOrAddress, nodes, network, segwit) {
    var scriptType = typeof pathOrAddress === 'string' ? isScriptHash(pathOrAddress, network) ? 'PAYTOSCRIPTHASH' : 'PAYTOADDRESS' : segwit ? 'PAYTOP2SHWITNESS' : 'PAYTOADDRESS';

    if (typeof pathOrAddress === 'string' && isBech32(pathOrAddress)) {
        var data = bitcoin.address.fromBech32(pathOrAddress).data;
        if (scriptType === 'PAYTOADDRESS') {
            return bitcoin.script.witnessPubKeyHash.output.encode(data);
        }

        if (scriptType === 'PAYTOSCRIPTHASH') {
            return bitcoin.script.witnessScriptHash.output.encode(data);
        }

        throw new Error('Unknown script type ' + scriptType);
    }

    var pkh = typeof pathOrAddress === 'string' ? bitcoin.address.fromBase58Check(pathOrAddress).hash : hdnodeUtils.derivePubKeyHash(nodes, pathOrAddress[pathOrAddress.length - 2], pathOrAddress[pathOrAddress.length - 1]);

    if (scriptType === 'PAYTOADDRESS') {
        return bitcoin.script.pubKeyHash.output.encode(pkh);
    }

    if (scriptType === 'PAYTOSCRIPTHASH') {
        return bitcoin.script.scriptHash.output.encode(pkh);
    }

    if (scriptType === 'PAYTOP2SHWITNESS') {
        return deriveWitnessOutput(pkh);
    }

    throw new Error('Unknown script type ' + scriptType);
}

function verifyBjsTx(inputs, outputs, nodes, resTx, network) {
    if (inputs.length !== resTx.ins.length) {
        throw new Error('Signed transaction has wrong length.');
    }
    if (outputs.length !== resTx.outs.length) {
        throw new Error('Signed transaction has wrong length.');
    }

    outputs.map(function (output, i) {
        var scriptB = resTx.outs[i].script;

        if (output.opReturnData instanceof Buffer) {
            var scriptA = bitcoin.script.nullData.output.encode(output.opReturnData);
            if (scriptA.compare(scriptB) !== 0) {
                throw new Error('Scripts differ');
            }
        } else {
            if (output.value !== resTx.outs[i].value) {
                throw new Error('Signed transaction has wrong output value.');
            }
            if (output.address === null && output.path === null) {
                throw new Error('Both path and address cannot be null.');
            }

            var addressOrPath = _flow_getPathOrAddress(output);
            var _segwit = _flow_getSegwit(output);
            var _scriptA = deriveOutputScript(addressOrPath, nodes, network, _segwit);
            if (_scriptA.compare(scriptB) !== 0) {
                throw new Error('Scripts differ');
            }
        }
    });
}

function isBech32(address) {
    try {
        bitcoin.address.fromBech32(address);
        return true;
    } catch (e) {
        return false;
    }
}

function isScriptHash(address, network, isCashaddress) {
    // cashaddr hack
    // Cashaddr format (with prefix) is neither base58 nor bech32, so it would fail
    // in trezor-utxo-lib. For this reason, we use legacy format here
    try {
        if (isCashaddress) {
            address = _bchaddrjs2.default.toLegacyAddress(address);
        }
    } catch (err) {
        throw new Error('Received cashaddr address could not be translated to legacy format for purpose of internal checks');
    }
    if (!isBech32(address)) {
        var decoded = bitcoin.address.fromBase58Check(address);
        if (decoded.version === network.pubKeyHash) {
            return false;
        }
        if (decoded.version === network.scriptHash) {
            return true;
        }
    } else {
        var _decoded = bitcoin.address.fromBech32(address);
        if (_decoded.data.length === 20) {
            return false;
        }
        if (_decoded.data.length === 32) {
            return true;
        }
    }
    throw new Error('Unknown address type.');
}

function signBjsTx(session, info, refTxs, nodes, coinName, network_, locktime, isCashaddress, overwintered) {
    var network = network_ == null ? bitcoin.networks[coinName.toLowerCase()] : network_;
    if (network == null) {
        return Promise.reject(new Error('No network ' + coinName));
    }

    // TODO rbf
    var sequence = locktime ? 0xffffffff - 1 : 0xffffffff;

    var trezorInputs = info.inputs.map(function (i) {
        return input2trezor(i, sequence);
    });
    // in case of bitcoin cash transaction, in output2trezor function actual conversion from legacy
    // to cashaddress format takes place
    var trezorOutputs = info.outputs.map(function (o) {
        return output2trezor(o, network, isCashaddress);
    });
    var trezorRefTxs = refTxs.map(function (tx) {
        return bjsTx2refTx(tx);
    });

    return session.signTx(trezorInputs, trezorOutputs, trezorRefTxs, coinName, locktime, overwintered).then(function (tx) {
        return signedTx2bjsTx(tx, network);
    }).then(function (res) {
        verifyBjsTx(info.inputs, info.outputs, nodes, res, network);
        return res;
    });
}

function reverseBuffer(buf) {
    var copy = new Buffer(buf.length);
    buf.copy(copy);
    [].reverse.call(copy);
    return copy;
}
